lattice_controller_engine/lib/lattice_controller_engine/observed/lattice.ex

```diff
@@ -0,0 +1,474 @@
+  end
+
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{
+            "public_key" => pk,
+            "instance_id" => instance_id,
+            "annotations" => %{
+              @annotation_app_spec => spec
+            }
+          },
+          datacontenttype: "application/json",
+          source: source_host,
+          type: "com.wasmcloud.lattice.actor_stopped"
+        }
+      ) do
+    remove_actor_instance(l, source_host, pk, instance_id, spec)
+  end
+
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{
+            "public_key" => pk,
+            "instance_id" => instance_id,
+            "annotations" => %{
+              @annotation_app_spec => spec
+            }
+          },
+          source: source_host,
+          datacontenttype: "application/json",
+          time: stamp,
+          type: "com.wasmcloud.lattice.actor_started"
+        }
+      ) do
+    put_actor_instance(l, source_host, pk, instance_id, spec, stamp)
+  end
+
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{
+            "actor_id" => actor_id,
+            "link_name" => link_name,
+            "contract_id" => contract_id,
+            "provider_id" => provider_id,
+            "values" => values
+          },
+          source: _source_host,
+          datacontenttype: "application/json",
+          type: "com.wasmcloud.lattice.linkdef_put"
+        }
+      ) do
+    put_linkdef(l, actor_id, link_name, provider_id, contract_id, values)
+  end
+
+  def apply_event(
+        l = %Lattice{},
+        %Cloudevents.Format.V_1_0.Event{
+          data: %{
+            "actor_id" => actor_id,
+            "link_name" => link_name,
+            "provider_id" => provider_id
+          },
+          source: _source_host,
+          datacontenttype: "application/json",
+          type: "com.wasmcloud.lattice.linkdef_del"
+        }
+      ) do
+    del_linkdef(l, actor_id, link_name, provider_id)
+  end
+
+  def apply_event(l = %Lattice{}, %Cloudevents.Format.V_1_0.Event{
+        data: %{
+          "image_ref" => image_ref,
+          "public_key" => public_key
+        },
+        source: _source_host,
+        datacontenttype: "application/json",
+        type: "com.wasmcloud.lattice.ocimap_put"
+      }) do
+    %Lattice{l | ocimap: Map.put(l.ocimap, image_ref, public_key)}
+  end
+
+  def apply_event(l = %Lattice{}, %Cloudevents.Format.V_1_0.Event{
+        data: %{
+          "image_ref" => image_ref
+        },
+        source: _source_host,
+        datacontenttype: "application/json",
+        type: "com.wasmcloud.lattice.ocimap_del"
+      }) do
+    %Lattice{l | ocimap: Map.delete(l.ocimap, image_ref)}
+  end
+
+  def apply_event(l = %Lattice{}, _evt = %Cloudevents.Format.V_1_0.Event{}) do
+    l
+  end
```

Comment:
I was going to comment to add a warn here for an unexpected event, but I actually like this so we don't get super spammy logs as we add more events that the controller doesn't care about 👍🏻